src/waveform/renderers/allshader/digitsrenderer.cpp

```diff
@@ -0,0 +1,196 @@
+        totalTextWidth += metrics.horizontalAdvance(text) + SPACE + SPACE;
+    }
+
+    totalTextWidth = std::ceil(totalTextWidth);
+
+    QImage image(static_cast<int>(totalTextWidth * devicePixelRatio),
+            static_cast<int>(std::ceil(maxTextHeight + SPACE + SPACE) * devicePixelRatio),
+            QImage::Format_ARGB32_Premultiplied);
+    image.setDevicePixelRatio(devicePixelRatio);
+    image.fill(Qt::transparent);
+
+    // Draw digits with dark outline
+    QPainter painter;
+    QPen pen(QColor(0, 0, 0, 192));
+    pen.setWidth(3);
+
+    painter.begin(&image);
+    painter.setRenderHint(QPainter::Antialiasing);
+    painter.setWorldMatrixEnabled(false);
+    painter.setBrush(QColor(0, 0, 0, 224));
+    painter.setPen(pen);
+    painter.setFont(font);
+    qreal x = 0;
+    for (int i = 0; i < NUM_CHARS; i++) {
+        const QString text(indexToChar(i));
+        QPainterPath path;
+        path.addText(QPointF(x + SPACE, maxTextHeight + SPACE), font, text);
+        painter.drawPath(path);
+        x += metrics.horizontalAdvance(text) + SPACE + SPACE;
+    }
+    painter.end();
+
+    // Apply Gaussian blur to dark outline
+
+    QGraphicsBlurEffect* blur = new QGraphicsBlurEffect; // ownership passed to item
+    blur->setBlurRadius(3);
+
+    QGraphicsScene scene;
+    QGraphicsPixmapItem item;
+    item.setPixmap(QPixmap::fromImage(image));
+    item.setGraphicsEffect(blur);
+    scene.addItem(&item);
+
+    image.fill(Qt::transparent);
+    painter.begin(&image);
+    scene.render(&painter, QRectF(), QRectF(0, 0, image.width(), image.height()));
+
+    // Draw digits foreground
+    painter.setPen(Qt::white);
+    painter.setFont(font);
+
+    x = 0;
+    for (int i = 0; i < NUM_CHARS; i++) {
+        const QString text(indexToChar(i));
+        painter.drawText(QPointF(x + SPACE, maxTextHeight + SPACE), text);
+        // position and width of character at index i in the texture
+        m_offset[i] = static_cast<float>(x / totalTextWidth);
+        m_width[i] = static_cast<float>(metrics.horizontalAdvance(text) + SPACE + SPACE);
+        x += metrics.horizontalAdvance(text) + SPACE + SPACE;
+    }
+    m_offset[NUM_CHARS] = 1.f;
+
+    painter.end();
+
+    m_texture.setData(image);
+}
+
+float allshader::DigitsRenderer::draw(const QMatrix4x4& matrix,
+        float x,
+        float y,
+        const QString& s,
+        float devicePixelRatio) {
+    const int n = s.length();
+    const float x0 = x;
+    const float dx = static_cast<float>(-(SPACE + SPACE) + KERNING);
+
+    VertexData posVertices;
+    VertexData texVertices;
+
+    posVertices.reserve(n * 6); // two triangles per character
+    texVertices.reserve(n * 6);
+
+    const QByteArray byteArray = s.toUtf8();
+
+    for (qsizetype i = 0; i < byteArray.size(); i++) {
+        // add vertices for the rectangles in the texture corresponding with each char in QString s
+        int index = charToIndex(byteArray[i]);
```

Comment:
why not index directly? The cast to UTF-8 along with indexing only works if you assume the string only contains ascii anyways (which is true yeah, but this is not a very safe way of expressing that).

```suggestion
    for (QChar c : s) {
        int index = charToIndex(c);
```